File: src/spectrl/formats/_peaklist_formats.py
"""MGF and MS2: peak lists with a small precursor header.

Both formats describe one fragmentation spectrum. They carry a precursor m/z,
usually a charge, sometimes a retention time, and then the peaks. Everything
else a token holds -- acquisition context, processing history, scan windows,
source files, additional arrays, units, extensions -- has nowhere to go, so the
writers report it rather than dropping it silently.

Neither format has a specification in the sense mzML does. MGF is defined by
what Mascot accepts and MS2 by what its readers parse, so these writers emit the
common subset that tools agree on and the readers accept the variations seen in
practice.
"""

from __future__ import annotations

import math
import re

from ..model import DecodedSpectrum, InlineSpectrum, SpectrlCvParam, SpectrlPrecursor, SpectrlScan, SpectrlSelectedIon
from ._report import ConversionResult

SELECTED_ION_MZ = "MS:1000744"
CHARGE_STATE = "MS:1000041"
PEAK_INTENSITY = "MS:1000042"
SCAN_START_TIME = "MS:1000016"
MS_LEVEL = "MS:1000511"
UO_SECOND = "UO:0000010"
UO_MINUTE = "UO:0000031"


def _find(params, accession):
    for param in params or []:
        if param.accession == accession:
            return param.value
    return None


def _precursor_of(spectrum):
    """The first selected ion's m/z, charge and intensity, if there is one."""
    for precursor in spectrum.precursors or []:
        for ion in precursor.selected_ions or []:
            mz = _find(ion.params, SELECTED_ION_MZ)
            if mz is not None:
                return (
                    float(mz),
                    _find(ion.params, CHARGE_STATE),
                    _find(ion.params, PEAK_INTENSITY),
                )
    return None


def _retention_seconds(spectrum):
    """Scan start time in seconds. mzML records minutes as often as seconds."""
    for scan in spectrum.scans or []:
        for param in scan.params or []:
            if param.accession == SCAN_START_TIME and param.value is not None:
                value = float(param.value)
                return value * 60 if param.unit_accession == UO_MINUTE else value
    return None


def _inventory(spectrum, result: ConversionResult) -> None:
    """Report what a precursor-plus-peaks format cannot represent."""
    checks = [
        (spectrum.acquisition, "acquisition", "acquisition context (instrument and components)"),
        (spectrum.processing, "processing", "processing history"),
        (spectrum.source, "source", "source file metadata"),
        (spectrum.products, "products", "product ion selection"),
        (spectrum.extensions, "extensions", "spectrum extensions"),
        (spectrum.array_extensions, "array_extensions", "array extensions"),
        (spectrum.cv_versions, "cv_versions", "ontology version provenance"),
        (spectrum.user_params, "user_params", "spectrum user parameters"),
    ]
    for value, path, description in checks:
        if value:
            result.add("dropped", path, description)
    extra = [k for k in (spectrum.extra_arrays or {})]
    if extra:
        result.add("dropped", "extra_arrays", f"additional arrays ({', '.join(sorted(extra))})")
    if spectrum.charge is not None:
        result.add("dropped", "charge", "per-peak charge array")
    if any(scan.windows for scan in spectrum.scans or []):
        result.add("dropped", "scans.windows", "scan windows")
    # Spectrum-level CV params beyond the few the header encodes.
    carried = {MS_LEVEL}
    extras = sorted({p.accession for p in spectrum.params or []} - carried)
    if extras:
        result.add("dropped", "params", f"{len(extras)} spectrum CV parameter(s), including {extras[0]}")


def _peaks(spectrum):
    mz = spectrum.mz if spectrum.mz is not None else []
    intensity = spectrum.intensity if spectrum.intensity is not None else []
    return zip(mz, intensity, strict=False)


def _number(value: float) -> str:
    """Round-trippable shortest form, without numpy's array scalar repr."""
    return repr(float(value))


def write_mgf(spectrum: DecodedSpectrum, *, title: str | None = None) -> ConversionResult:
    """Write one spectrum as a single-entry MGF.

    A spectrum with no precursor is written without PEPMASS and reported,
    because MGF's own convention expects one and many readers require it.
    """
    result = ConversionResult("", "mgf")
    _inventory(spectrum, result)
    precursor = _precursor_of(spectrum)
    lines = ["BEGIN IONS", f"TITLE={title or spectrum.id or 'spectrl'}"]
    if precursor is None:
        result.add(
            "no_precursor",
            "precursors",
            "no precursor ion, so no PEPMASS was written; most MGF readers expect one",
        )
    else:
        mz, charge, intensity = precursor
        lines.append(f"PEPMASS={_number(mz)}" + (f" {_number(float(intensity))}" if intensity is not None else ""))
        if charge is not None:
            value = int(float(charge))
            lines.append(f"CHARGE={abs(value)}{'+' if value >= 0 else '-'}")
    seconds = _retention_seconds(spectrum)
    if seconds is not None:
        lines.append(f"RTINSECONDS={_number(seconds)}")
    for mz, intensity in _peaks(spectrum):
        lines.append(f"{_number(mz)} {_number(intensity)}")
    lines.append("END IONS")
    result.text = "\n".join(lines) + "\n"
    return result


def write_ms2(spectrum: DecodedSpectrum, *, scan: int | None = None) -> ConversionResult:
    """Write one spectrum as a single-entry MS2 file."""
    result = ConversionResult("", "ms2")
    _inventory(spectrum, result)
    precursor = _precursor_of(spectrum)
    number = scan if scan is not None else _scan_number(spectrum.id)
    lines = [
        "H\tCreationDate\t",
        "H\tExtractor\tspectrl",
    ]
    if precursor is None:
        result.add(
            "no_precursor",
            "precursors",
            "no precursor ion; the MS2 S line requires a precursor m/z, so 0 was written",
        )
        lines.append(f"S\t{number}\t{number}\t0")
    else:
        mz, charge, _ = precursor
        lines.append(f"S\t{number}\t{number}\t{_number(mz)}")
        seconds = _retention_seconds(spectrum)
        if seconds is not None:
            lines.append(f"I\tRTime\t{_number(seconds / 60)}")
        if charge is not None:
            state = abs(int(float(charge))) or 1
            # The Z line's mass is the singly-protonated neutral mass.
            lines.append(f"Z\t{state}\t{_number(mz * state - (state - 1) * 1.00727646677)}")
    for mz, intensity in _peaks(spectrum):
        lines.append(f"{_number(mz)} {_number(intensity)}")
    result.text = "\n".join(lines) + "\n"
    return result


def _scan_number(spectrum_id: str | None) -> int:
    """A scan number for the S line, from a native id when one is present."""
    if spectrum_id:
        match = re.search(r"scan=(\d+)", spectrum_id) or re.search(r"(\d+)\s*$", spectrum_id)
        if match:
            return int(match.group(1))
    return 0


def _spectrum(mz, intensity, *, spectrum_id, ms_level, precursor, retention_seconds):
    """Build an InlineSpectrum from what a peak-list format supplies."""
    import numpy as np  # noqa: PLC0415

    params = [SpectrlCvParam(MS_LEVEL, ms_level)]
    scans = []
    if retention_seconds is not None:
        scans = [SpectrlScan(params=[SpectrlCvParam(SCAN_START_TIME, retention_seconds, UO_SECOND)])]
    precursors = []
    if precursor is not None:
        mz_value, charge, intensity_value = precursor
        ion = [SpectrlCvParam(SELECTED_ION_MZ, mz_value)]
        if charge is not None:
            ion.append(SpectrlCvParam(CHARGE_STATE, charge))
        if intensity_value is not None:
            ion.append(SpectrlCvParam(PEAK_INTENSITY, intensity_value))
        precursors = [SpectrlPrecursor(selected_ions=[SpectrlSelectedIon(params=ion)])]
    return InlineSpectrum(
        default_array_length=len(mz),
        mz=np.asarray(mz, dtype="float64"),
        intensity=np.asarray(intensity, dtype="float64"),
        id=spectrum_id,
        params=params,
        scans=scans,
        precursors=precursors,
    )


def _peak_line(line: str, number: int, where: str):
    parts = line.split()
    if len(parts) < 2:
        raise ValueError(f"{where} line {number}: expected 'mz intensity', got {line!r}")
    try:
        mz, intensity = float(parts[0]), float(parts[1])
    except ValueError:
        raise ValueError(f"{where} line {number}: peak values must be numbers, got {line!r}") from None
    if not (math.isfinite(mz) and math.isfinite(intensity)):
        raise ValueError(f"{where} line {number}: peak values must be finite")
    return mz, intensity


_CHARGE = re.compile(r"^([+-]?\d+)([+-]?)$")


def _parse_charge(text: str) -> int | None:
    """MGF writes 2+, +2, or 2; a list like '2+ and 3+' is genuinely ambiguous."""
    token = text.strip().split()[0] if text.strip() else ""
    match = _CHARGE.match(token)
    if not match:
        return None
    value = int(match.group(1))
    return -abs(value) if match.group(2) == "-" else abs(value)


def read_mgf(text: str) -> list[InlineSpectrum]:
    """Parse every BEGIN IONS block in an MGF document."""
    spectra: list[InlineSpectrum] = []
    header: dict[str, str] = {}
    mz: list[float] = []
    intensity: list[float] = []
    inside = False
    for number, raw in enumerate(text.splitlines(), start=1):
        line = raw.strip()
        if not line or line.startswith(("#", ";", "!")):
            continue
        upper = line.upper()
        if upper == "BEGIN IONS":
            if inside:
                raise ValueError(f"MGF line {number}: BEGIN IONS inside an open block")
            inside, header, mz, intensity = True, {}, [], []
            continue
        if upper == "END IONS":
            if not inside:
                raise ValueError(f"MGF line {number}: END IONS without BEGIN IONS")
            spectra.append(_from_mgf_block(header, mz, intensity))
            inside = False
            continue
        if not inside:
            continue  # Global parameters before the first block apply to search, not to peaks.
        if "=" in line and not line[0].isdigit():
            key, _, value = line.partition("=")
            header[key.strip().upper()] = value.strip()
            continue
        a, b = _peak_line(line, number, "MGF")
        mz.append(a)
        intensity.append(b)
    if inside:
        raise ValueError("MGF ended inside an unterminated BEGIN IONS block")
    return spectra


def _from_mgf_block(header, mz, intensity):
    precursor = None
    if "PEPMASS" in header:
        parts = header["PEPMASS"].split()
        precursor = (
            float(parts[0]),
            _parse_charge(header["CHARGE"]) if "CHARGE" in header else None,
            float(parts[1]) if len(parts) > 1 else None,
        )
    seconds = None
    if "RTINSECONDS" in header:
        seconds = float(header["RTINSECONDS"].split()[0])
    return _spectrum(
        mz,
        intensity,
        spectrum_id=header.get("TITLE"),
        # MGF describes fragmentation, so a block without a precursor is still
        # read as MS2 rather than guessed at.
        ms_level=2,
        precursor=precursor,
        retention_seconds=seconds,
    )


_PROTON = 1.00727646677


def read_ms2(text: str) -> list[InlineSpectrum]:
    """Parse every S block in an MS2 document."""
    spectra: list[InlineSpectrum] = []
    current: dict | None = None
    mz: list[float] = []
    intensity: list[float] = []

    def flush():
        if current is not None:
            spectra.append(
                _spectrum(
                    mz,
                    intensity,
                    spectrum_id=f"scan={current['scan']}" if current["scan"] else None,
                    ms_level=2,
                    precursor=current["precursor"],
                    retention_seconds=current["rtime"],
                )
            )

    for number, raw in enumerate(text.splitlines(), start=1):
        line = raw.rstrip()
        if not line:
            continue
        kind = line[0]
        if kind == "H":
            continue
        if kind == "S":
            flush()
            parts = line.split()
            if len(parts) < 4:
                raise ValueError(f"MS2 line {number}: S line needs scan, scan and precursor m/z")
            precursor_mz = float(parts[3])
            current = {
                "scan": parts[1],
                "precursor": (precursor_mz, None, None) if precursor_mz else None,
                "rtime": None,
            }
            mz, intensity = [], []
            continue
        if current is None:
            continue
        if kind == "I":
            parts = line.split()
            if len(parts) >= 3 and parts[1].upper() == "RTIME":
                current["rtime"] = float(parts[2]) * 60  # MS2 records RTime in minutes.
            continue
        if kind == "Z":
            parts = line.split()
            if len(parts) >= 2 and current["precursor"]:
                state = int(float(parts[1]))
                mz_value, _, intensity_value = current["precursor"]
                current["precursor"] = (mz_value, state, intensity_value)
            continue
        if kind == "D":
            continue
        a, b = _peak_line(line, number, "MS2")
        mz.append(a)
        intensity.append(b)
    flush()
    return spectra

File: js/src/formats/peaklist_formats.ts
/** MGF and MS2: peak lists with a small precursor header.
 *
 * Both describe one fragmentation spectrum -- a precursor m/z, usually a
 * charge, sometimes a retention time, then peaks. Everything else a token holds
 * has nowhere to go, so the writers report it. Neither format has a
 * specification in the sense mzML does: MGF is what Mascot accepts and MS2 is
 * what its readers parse, so these emit the common subset tools agree on and
 * accept the variations seen in practice.
 *
 * Mirrors src/spectrl/formats/_peaklist_formats.py. */

import type { CvParam, DecodedSpectrum, InlineSpectrum } from "../model.js"
import { ConversionResult } from "./report.js"

export const SELECTED_ION_MZ = "MS:1000744"
export const CHARGE_STATE = "MS:1000041"
export const PEAK_INTENSITY = "MS:1000042"
export const SCAN_START_TIME = "MS:1000016"
export const MS_LEVEL = "MS:1000511"
const UO_SECOND = "UO:0000010"
const UO_MINUTE = "UO:0000031"
const PROTON = 1.00727646677

type Precursor = { mz: number, charge: number | null, intensity: number | null }

function find(params: CvParam[] | undefined, accession: string): unknown {
  for (const p of params ?? []) if (p.accession === accession) return p.value
  return null
}

function precursorOf(spectrum: DecodedSpectrum): Precursor | null {
  for (const precursor of spectrum.precursors ?? []) {
    for (const ion of precursor.selectedIons ?? []) {
      const mz = find(ion.params, SELECTED_ION_MZ)
      if (mz != null) {
        const charge = find(ion.params, CHARGE_STATE)
        const intensity = find(ion.params, PEAK_INTENSITY)
        return {
          mz: Number(mz),
          charge: charge == null ? null : Number(charge),
          intensity: intensity == null ? null : Number(intensity),
        }
      }
    }
  }
  return null
}

/** Scan start time in seconds; mzML records minutes as often as seconds. */
function retentionSeconds(spectrum: DecodedSpectrum): number | null {
  for (const scan of spectrum.scans ?? []) {
    for (const p of scan.params ?? []) {
      if (p.accession === SCAN_START_TIME && p.value != null) {
        const value = Number(p.value)
        return p.unitAccession === UO_MINUTE ? value * 60 : value
      }
    }
  }
  return null
}

/** Report what a precursor-plus-peaks format cannot represent. */
function inventory(spectrum: DecodedSpectrum, result: ConversionResult): void {
  const checks: [unknown, string, string][] = [
    [spectrum.acquisition, "acquisition", "acquisition context (instrument and components)"],
    [spectrum.processing, "processing", "processing history"],
    [spectrum.source, "source", "source file metadata"],
    [spectrum.products, "products", "product ion selection"],
    [spectrum.extensions, "extensions", "spectrum extensions"],
    [spectrum.arrayExtensions, "array_extensions", "array extensions"],
    [spectrum.cvVersions, "cv_versions", "ontology version provenance"],
    [spectrum.userParams, "user_params", "spectrum user parameters"],
  ]
  for (const [value, path, description] of checks) {
    const present = Array.isArray(value) ? value.length > 0 : value && Object.keys(value).length > 0
    if (present) result.add("dropped", path, description)
  }
  const extra = Object.keys(spectrum.extraArrays ?? {})
  if (extra.length) result.add("dropped", "extra_arrays", `additional arrays (${extra.sort().join(", ")})`)
  if (spectrum.charge != null) result.add("dropped", "charge", "per-peak charge array")
  if ((spectrum.scans ?? []).some(s => s.windows?.length)) result.add("dropped", "scans.windows", "scan windows")
  const extras = [...new Set((spectrum.params ?? []).map(p => p.accession))].filter(a => a !== MS_LEVEL).sort()
  if (extras.length) {
    result.add("dropped", "params", `${extras.length} spectrum CV parameter(s), including ${extras[0]}`)
  }
}

/** Shortest round-trippable form, matching CPython's repr(float).
 *
 * Both languages print the shortest digits that round-trip, but they disagree
 * on when to switch to scientific notation and how to spell the exponent.
 * CPython goes exponential once the decimal point would sit at or before the
 * fourth leading zero, or past the sixteenth digit, and always writes at least
 * two exponent digits: 1e-05, not 0.00001; 4.5e-07, not 4.5e-7. JavaScript
 * switches at 1e-7 and pads nothing, so the two writers would otherwise emit
 * different text for the same value. */
export function num(value: number): string {
  if (!Number.isFinite(value)) throw Error(`cannot write a non-finite value: ${value}`)
  if (value === 0) return Object.is(value, -0) ? "-0.0" : "0.0"
  const parsed = /^(-?)(\d)(?:\.(\d+))?e([+-]\d+)$/.exec(value.toExponential())
  if (!parsed) throw Error(`unexpected number form: ${value}`)
  const [, sign, first, rest = "", exponentText] = parsed
  const exponent = Number(exponentText)
  const digits = first! + rest
  if (exponent <= -5 || exponent >= 16) {
    const mantissa = rest ? `${first}.${rest}` : first!
    const magnitude = String(Math.abs(exponent)).padStart(2, "0")
    return `${sign}${mantissa}e${exponent < 0 ? "-" : "+"}${magnitude}`
  }
  if (exponent >= 0) {
    if (digits.length <= exponent + 1) return `${sign}${digits.padEnd(exponent + 1, "0")}.0`
    return `${sign}${digits.slice(0, exponent + 1)}.${digits.slice(exponent + 1)}`
  }
  return `${sign}0.${"0".repeat(-exponent - 1)}${digits}`
}

function* peaks(spectrum: DecodedSpectrum): Generator<[number, number]> {
  const mz = spectrum.mz ?? []
  const intensity = spectrum.intensity ?? []
  const n = Math.min(mz.length, intensity.length)
  for (let i = 0; i < n; i++) yield [mz[i]!, intensity[i]!]
}

export function writeMgf(spectrum: DecodedSpectrum, opts: { title?: string } = {}): ConversionResult {
  const result = new ConversionResult("mgf")
  inventory(spectrum, result)
  const precursor = precursorOf(spectrum)
  const lines = ["BEGIN IONS", `TITLE=${opts.title || spectrum.id || "spectrl"}`]
  if (precursor === null) {
    result.add("no_precursor", "precursors",
      "no precursor ion, so no PEPMASS was written; most MGF readers expect one")
  } else {
    lines.push(`PEPMASS=${num(precursor.mz)}` + (precursor.intensity != null ? ` ${num(precursor.intensity)}` : ""))
    if (precursor.charge != null) {
      lines.push(`CHARGE=${Math.abs(precursor.charge)}${precursor.charge >= 0 ? "+" : "-"}`)
    }
  }
  const seconds = retentionSeconds(spectrum)
  if (seconds != null) lines.push(`RTINSECONDS=${num(seconds)}`)
  for (const [mz, intensity] of peaks(spectrum)) lines.push(`${num(mz)} ${num(intensity)}`)
  lines.push("END IONS")
  result.text = lines.join("\n") + "\n"
  return result
}

function scanNumber(id: string | null): number {
  if (id) {
    const match = /scan=(\d+)/.exec(id) ?? /(\d+)\s*$/.exec(id)
    if (match) return Number(match[1])
  }
  return 0
}

export function writeMs2(spectrum: DecodedSpectrum, opts: { scan?: number } = {}): ConversionResult {
  const result = new ConversionResult("ms2")
  inventory(spectrum, result)
  const precursor = precursorOf(spectrum)
  const number = opts.scan ?? scanNumber(spectrum.id)
  const lines = ["H\tCreationDate\t", "H\tExtractor\tspectrl"]
  if (precursor === null) {
    result.add("no_precursor", "precursors",
      "no precursor ion; the MS2 S line requires a precursor m/z, so 0 was written")
    lines.push(`S\t${number}\t${number}\t0`)
  } else {
    lines.push(`S\t${number}\t${number}\t${num(precursor.mz)}`)
    const seconds = retentionSeconds(spectrum)
    if (seconds != null) lines.push(`I\tRTime\t${num(seconds / 60)}`)
    if (precursor.charge != null) {
      const state = Math.abs(precursor.charge) || 1
      // The Z line's mass is the singly-protonated neutral mass.
      lines.push(`Z\t${state}\t${num(precursor.mz * state - (state - 1) * PROTON)}`)
    }
  }
  for (const [mz, intensity] of peaks(spectrum)) lines.push(`${num(mz)} ${num(intensity)}`)
  result.text = lines.join("\n") + "\n"
  return result
}

function spectrumFrom(
  mz: number[], intensity: number[],
  o: { id?: string | null, msLevel: number, precursor: Precursor | null, retentionSeconds: number | null },
): InlineSpectrum {
  const params: CvParam[] = [{ accession: MS_LEVEL, value: o.msLevel }]
  const scans = o.retentionSeconds == null ? [] : [{
    params: [{ accession: SCAN_START_TIME, value: o.retentionSeconds, unitAccession: UO_SECOND }],
    windows: [], userParams: [],
  }]
  const precursors = []
  if (o.precursor) {
    const ion: CvParam[] = [{ accession: SELECTED_ION_MZ, value: o.precursor.mz }]
    if (o.precursor.charge != null) ion.push({ accession: CHARGE_STATE, value: o.precursor.charge })
    if (o.precursor.intensity != null) ion.push({ accession: PEAK_INTENSITY, value: o.precursor.intensity })
    precursors.push({
      isolationWindow: null, activation: null,
      selectedIons: [{ params: ion, userParams: [] }],
    })
  }
  return {
    defaultArrayLength: mz.length,
    mz: Float64Array.from(mz),
    intensity: Float64Array.from(intensity),
    id: o.id ?? null,
    params, scans, precursors,
  } as InlineSpectrum
}

function peakLine(line: string, number: number, where: string): [number, number] {
  const parts = line.split(/\s+/).filter(Boolean)
  if (parts.length < 2) throw Error(`${where} line ${number}: expected 'mz intensity', got ${JSON.stringify(line)}`)
  const mz = Number(parts[0]), intensity = Number(parts[1])
  if (!Number.isFinite(mz) || !Number.isFinite(intensity)) {
    throw Error(`${where} line ${number}: peak values must be finite numbers, got ${JSON.stringify(line)}`)
  }
  return [mz, intensity]
}

/** MGF writes 2+, +2, or 2; a list like "2+ and 3+" is genuinely ambiguous. */
function parseCharge(text: string): number | null {
  const token = text.trim().split(/\s+/)[0] ?? ""
  const match = /^([+-]?\d+)([+-]?)$/.exec(token)
  if (!match) return null
  const value = Math.abs(Number(match[1]))
  return match[2] === "-" ? -value : value
}

export function readMgf(text: string): InlineSpectrum[] {
  const spectra: InlineSpectrum[] = []
  let header: Record<string, string> = {}
  let mz: number[] = [], intensity: number[] = []
  let inside = false
  let number = 0
  for (const raw of text.split(/\r\n|\r|\n/)) {
    number++
    const line = raw.trim()
    if (!line || "#;!".includes(line[0]!)) continue
    const upper = line.toUpperCase()
    if (upper === "BEGIN IONS") {
      if (inside) throw Error(`MGF line ${number}: BEGIN IONS inside an open block`)
      inside = true; header = {}; mz = []; intensity = []
      continue
    }
    if (upper === "END IONS") {
      if (!inside) throw Error(`MGF line ${number}: END IONS without BEGIN IONS`)
      spectra.push(fromMgfBlock(header, mz, intensity))
      inside = false
      continue
    }
    if (!inside) continue // Parameters before the first block configure search, not peaks.
    if (line.includes("=") && !/^\d/.test(line)) {
      const at = line.indexOf("=")
      header[line.slice(0, at).trim().toUpperCase()] = line.slice(at + 1).trim()
      continue
    }
    const [a, b] = peakLine(line, number, "MGF")
    mz.push(a); intensity.push(b)
  }
  if (inside) throw Error("MGF ended inside an unterminated BEGIN IONS block")
  return spectra
}

function fromMgfBlock(header: Record<string, string>, mz: number[], intensity: number[]): InlineSpectrum {
  let precursor: Precursor | null = null
  if (header.PEPMASS) {
    const parts = header.PEPMASS.split(/\s+/).filter(Boolean)
    precursor = {
      mz: Number(parts[0]),
      charge: header.CHARGE ? parseCharge(header.CHARGE) : null,
      intensity: parts.length > 1 ? Number(parts[1]) : null,
    }
  }
  const rt = header.RTINSECONDS ? Number(header.RTINSECONDS.split(/\s+/)[0]) : null
  // MGF describes fragmentation, so a block without a precursor is still read
  // as MS2 rather than guessed at.
  return spectrumFrom(mz, intensity, { id: header.TITLE ?? null, msLevel: 2, precursor, retentionSeconds: rt })
}

export function readMs2(text: string): InlineSpectrum[] {
  const spectra: InlineSpectrum[] = []
  let current: { scan: string, precursor: Precursor | null, rtime: number | null } | null = null
  let mz: number[] = [], intensity: number[] = []
  let number = 0
  const flush = () => {
    if (current) {
      spectra.push(spectrumFrom(mz, intensity, {
        id: current.scan ? `scan=${current.scan}` : null,
        msLevel: 2, precursor: current.precursor, retentionSeconds: current.rtime,
      }))
    }
  }
  for (const raw of text.split(/\r\n|\r|\n/)) {
    number++
    const line = raw.replace(/\s+$/, "")
    if (!line) continue
    const kind = line[0]!
    if (kind === "H") continue
    if (kind === "S") {
      flush()
      const parts = line.split(/\s+/).filter(Boolean)
      if (parts.length < 4) throw Error(`MS2 line ${number}: S line needs scan, scan and precursor m/z`)
      const precursorMz = Number(parts[3])
      if (Number.isNaN(precursorMz)) throw Error(`MS2 line ${number}: precursor m/z is not a number: ${JSON.stringify(parts[3])}`)
      current = { scan: parts[1]!, precursor: precursorMz ? { mz: precursorMz, charge: null, intensity: null } : null, rtime: null }
      mz = []; intensity = []
      continue
    }
    if (!current) continue
    if (kind === "I") {
      const parts = line.split(/\s+/).filter(Boolean)
      if (parts.length >= 3 && parts[1]!.toUpperCase() === "RTIME") current.rtime = Number(parts[2]) * 60
      continue
    }
    if (kind === "Z") {
      const parts = line.split(/\s+/).filter(Boolean)
      if (parts.length >= 2 && current.precursor) current.precursor.charge = Math.trunc(Number(parts[1]))
      continue
    }
    if (kind === "D") continue
    const [a, b] = peakLine(line, number, "MS2")
    mz.push(a); intensity.push(b)
  }
  flush()
  return spectra
}
